File: android/app/src/main/java/com/connect/club/react/bridge/iosLike/ui/AppManager.kt
package com.connect.club.react.bridge.iosLike.ui

import android.view.View
import androidx.core.view.isVisible
import com.connect.club.react.AppInfoModule
import com.connect.club.react.bridge.iosLike.NewJitsi
import com.connect.club.react.bridge.iosLike.debugP
import com.connect.club.react.bridge.iosLike.weak
import com.connect.club.react.rncviews.rncroomlayout.RNCFloatingReactionsView
import com.connect.club.react.rncviews.rncroomlisteners.ListenersPresenter
import com.connect.club.utils.forEach
import com.connect.club.utils.runOnMainThread
import common.LiveRoom
import org.json.JSONArray
import org.json.JSONObject
import java.util.*

object AppManager {

    val speakers = mutableMapOf<String, SpeakerView>()
    var targetPathView by weak<TargetPathView?>()
    var radarView by weak<RadarView?>()
    var shareScreenVideoView by weak<ShareDesktopContainerView?>()
    var reactionsView by weak<RNCFloatingReactionsView?>()

    var toggleVideoAudioButtonsSet = ToggleVideoAudioButtonsSet()
    var newJitsi by weak<NewJitsi?>()

    var room: LiveRoom? = null
    var roomName: String = ""
    var shouldUpdateUserPositionOnState = true

    var localUserId: String = ""

    @Volatile
    var isDestroying = false

    @Volatile
    var isJitsiConnected = false

    var isOffline = false

    val currentUserState: JSONObject?
        get() = synchronized(speakers) { speakers[localUserId]?.lastState }

    val currentSpeaker: SpeakerView?
        get() = synchronized(speakers) { speakers[localUserId] }


    fun addSpeaker(id: String, view: SpeakerView) {
        debugP("||| add speaker", id)
        speakers[id] = view
    }

    fun removeSpeaker(id: String) {
        synchronized(speakers) { speakers.remove(id) }
        room?.removeReaction(id)
    }

    fun sendMove(x: Float, y: Float) {
        debugP("View: sendMove", x, y)
        targetPathView?.move(x, y)
        room?.sendUserPath(x.toDouble(), y.toDouble())
    }

    fun moveUserTo(user: View, x: Double, y: Double, duration: Double, userId: String) {
        // wait radar only for current user
        if (radarView == null && userId == localUserId) return
        val radar = radarView
        val x = x.toFloat()
        val y = y.toFloat()

        val halfWidth = user.width / 2
        val halfHeight = user.height / 2
        val tx = user.translationX + halfWidth
        val ty = user.translationY + halfHeight

        if (x.toInt() == tx.toInt() && y.toInt() == ty.toInt()) return
        debugP("View: moveUser", userId, x, y, user.translationX, user.translationY, user.width)
        if (x.toInt() == 0 || y.toInt() == 0) {
            val speaker = user as? SpeakerView
            speaker?.move(0f, 0f)
            speaker?.isVisible = false
            if (userId == localUserId) {
                radarView?.isVisible = false
                targetPathView?.isVisible = false
            }
            return
        }

        if (userId == localUserId) targetPathView?.move(x, y)

        val userView = (user as BaseMovableView)
        if (userView.translationX < 0) userView.move(x, y)
        else userView.animatedMove(x, y, duration)

        if (userId == localUserId) {
            val rtx = radar?.translationX ?: 0F
            if (rtx == -10000f) radar?.move(x, y)
            else radar?.animatedMove(x, y, duration)
        }
    }

    fun moveUser(userId: String, x: Double, y: Double) {
        if (isDestroying) return
        val view = synchronized(speakers) { speakers[userId] } ?: return
        val isFirstMove = view.translationX <= 0 || view.translationY <= 0
        if (isFirstMove || shouldUpdateUserPositionOnState) {
            debugP("++++++ move user")
            moveUserTo(view, x, y, 1.0, userId)
        }
    }

    fun moveUserOnUI(userId: String, x: Double, y: Double, duration: Double) {
        runOnMainThread {
            if (isDestroying) return@runOnMainThread
            val view = synchronized(speakers) { speakers[userId] } ?: return@runOnMainThread
            moveUserTo(view, x, y, duration, userId)
        }
    }

    fun onNativeState(json: ByteArray?) {
        json ?: return
        val string = String(json)
        debugP("onNativeState", string)
        val array = JSONArray(string)
        runOnMainThread {
            array.forEach { state ->
                val id = state.getString("id")
                val x = state.getDouble("x")
                val y = state.getDouble("y")
                moveUser(id, x, y)
                if (isDestroying) return@runOnMainThread
                synchronized(speakers) { speakers[id] }?.let {
                    val scale = state.getDouble("audioLevel")
                    it.animateScale(scale.toFloat())
                    it.onState(state)
                    if (localUserId == id) toggleVideoAudioButtonsSet.onState(state)
                }
            }
            shouldUpdateUserPositionOnState = false
        }
    }

    fun onReaction(json: ByteArray?) {
        json ?: return
        val string = String(json)
        val jsonObject = JSONObject(string)
        runOnMainThread {
            val payload = jsonObject.getJSONObject("payload")
            val userId = payload.getString("fromId")
            val reaction = payload.getString("reaction")
            speakers[userId]?.toggleReaction(userId, reaction)
            ListenersPresenter.updateReaction(userId, reaction)
        }
    }

    fun onConnectionState(state: ByteArray?) {
        state ?: return
        val jsonString = String(state)
        val json = JSONObject(jsonString)
        if (json.getString("id") != localUserId) return
        runOnMainThread {
            toggleVideoAudioButtonsSet.isVisible = false
            if (json.getString("mode") != "popup") return@runOnMainThread
            synchronized(speakers) {
                val speaker = speakers.remove(localUserId) ?: return@synchronized
                moveUserTo(speaker, 0.0, 0.0, 1.0, localUserId)
            }
        }
    }
}

File: ios/Support/AppManager.swift
//
//  AppManager.swift
//  connectreactive
//
//  Created by Sergei Golishnikov on 10/04/2021.
//

import Foundation
import Common
import JitsiWebRTC

final class WeakSpeakerRef {
    weak var value: SpeakerView?
    init(value: SpeakerView) {
        self.value = value
    }
}

final class AppManager {
    static let instance = AppManager()
    
    static var isJitsiConnected = false
    
    static let isDestroyingSync = DispatchQueue(label: "isDestroying")
    static var isOffline = false
    
    private static var _isDestroying = false
    
    static var isDestroying: Bool {
        isDestroyingSync.sync {
            return _isDestroying
        }
    }
    
    static func setDestroying(_ isDestroying: Bool) {
        isDestroyingSync.sync {
            debugLog(.debug, isDestroying)
            _isDestroying = isDestroying
        }
    }
    
    var speakers = [String: WeakSpeakerRef]()
    weak var shareScreenVideoView: ShareDesktopContainerView?
    var toggleVideoAudioButtons: Set<ToggleVideoAudioButtons> = []
    
    weak var targetPathView: TargetPathView?
    weak var radarView: RadarView?
    var localUserId: String = ""
    var roomName: String = ""
    
    weak var newJitsi: NewJitsi?
    var room: CommonLiveRoom?
    
    var shouldUpdateUserPositionOnState = true
    
    static var speakers: [String: WeakSpeakerRef] {
        return instance.speakers
    }
    
    static var currentSpeaker: SpeakerView? {
        return speakers[instance.localUserId]?.value
    }
    
    static var localUserId: String { instance.localUserId }
    
    var onNewSpeaker: ((WeakSpeakerRef) -> Void)?
    
    func addSpeaker(_ id: String, _ view: SpeakerView) {
        let weakRef = WeakSpeakerRef(value: view)
        speakers[id] = weakRef
        
        onNewSpeaker?(weakRef)
    }
    
    func removeSpeaker(_ id: String) {
        speakers.removeValue(forKey: id)
        room?.removeReaction(id)
    }
    
    func sendMove(x: CGFloat, y: CGFloat) {
        debugLog(.info, x, y)
        targetPathView?.move(x, y)
        room?.sendUserPath(Double(x), toY: Double(y))
    }
    
    func moveUserTo(
        _ user: UIView,
        _ x: CGFloat,
        _ y: CGFloat,
        _ duration: Double,
        _ userId: String
    ) {
        // wait radar only for current user
        if radarView == nil && userId == localUserId { return }
        if userId == localUserId {
            targetPathView?.move(x, y)
        }
        
        let halfWidth = user.frame.width / 2
        let halfHeight = user.frame.width / 2
        let tx = user.frame.origin.x + halfWidth
        let ty = user.frame.origin.y + halfHeight
        if Int(tx) == Int(x) && Int(ty) == Int(y)  { return }
        
        if x == 0 || y == 0 {
            let speaker = (user as? SpeakerView)
            speaker?.move(x, y)
            speaker?.isHidden = true
            if userId == Self.instance.localUserId {
                Self.instance.radarView?.move(x, y)
                Self.instance.radarView?.isHidden = true
                Self.instance.targetPathView?.isHidden = true
            }
            return
        }

        debugLog(.debug, "id=\(userId) x=\(x) y=\(y) t=\(duration)")
        UIView.animate(withDuration: user.frame.origin.x <= 0 ? 0 : duration) {
            (user as? SpeakerView)?.move(x, y)
            if userId == Self.instance.localUserId {
                Self.instance.radarView?.move(x, y)
            }
        }
    }
    
    func moveUser(
        _ userId: String,
        _ x: Double,
        _ y: Double
    ) {
        if AppManager.isDestroying { return }
        guard let view = Self.instance.speakers[userId]?.value else { return }
        let isFirstPosition = view.frame.origin.x <= 0 || view.frame.origin.y <= 0
        if shouldUpdateUserPositionOnState || isFirstPosition {
            debugLog(.debug, "initial move id=\(userId) x=\(x) y=\(y)")
            moveUserTo(view, CGFloat(x), CGFloat(y), 1, userId)
        }
    }
    
    func moveUserOnUI(
        _ userId: String,
        _ x: Double,
        _ y: Double,
        _ duration: Double
    ) {
        DispatchQueue.main.async {
            if AppManager.isDestroying { return }
            guard let view = Self.instance.speakers[userId]?.value else { return }
            Self.instance.moveUserTo(view, CGFloat(x), CGFloat(y), duration, userId)
        }
    }
    
    func onNativeCurrentUserState(
        _ userId: String?,
        x: Double,
        y: Double,
        scale: Double
    ) {
        guard let userId = userId else { return }
        guard let view = Self.instance.speakers[userId]?.value else { return }
        moveUserOnUI(userId, x, y, 1)
        DispatchQueue.main.async {
            view.animateScale(scale)
        }
    }
    
    func onNativeState(_ json: Data?) {
        let jsonDict = dataToJSON(data: json) as! [[String: Any]]
        DispatchQueue.main.async {
            for entry in jsonDict {
                let i = AppManager.instance
                let model = JSON(entry)
                let id = model.id.string!
                i.moveUser(
                    id,
                    model.x.double!,
                    model.y.double!
                )
                if AppManager.isDestroying { return }
                if let view = i.speakers[id]?.value {
                    if let scale = model.audioLevel.double { view.animateScale(scale) }
                    view.onState(state: model)
                }
                
                if id == Self.instance.localUserId {
                    Self.instance.toggleVideoAudioButtons.forEach{ $0.onState(model) }
                }
            }
            self.shouldUpdateUserPositionOnState = false
        }
    }
    
    func onReaction(_ json: Data?) {
        guard let json = json else { return }
        guard let reaction = try? JSON(data: json) else { return }
        DispatchQueue.main.async { [weak self] in
            let id = reaction.payload.fromId.string!
            let type = reaction.payload.reaction.string!
            
            if let view = self?.speakers[id] {
                view.value?.toggleReaction(id: id, type)
            } else {
                ListenersStore.shared.onNewReaction?(id, type)
            }
        }
    }
    
    func onConnectionState(state: Data?) {
        guard let state = state else { return }
        guard let json = try? JSON(data: state) else { return }
        if json.id.string != localUserId { return }
        
        DispatchQueue.main.async {
            self.toggleVideoAudioButtons.forEach{ $0.isVisible = false }
            if json.mode.string != "popup" { return }
            guard let speaker = Self.speakers[self.localUserId]?.value else {
                return
            }
            self.moveUserTo(speaker, 0, 0, 1, self.localUserId)
        }
    }
    
    func sendViewPort(_ rect: CGRect) {
        room?.setViewport(
            Double(rect.origin.x),
            y1: Double(rect.origin.y),
            x2: Double(rect.maxX),
            y2: Double(rect.maxY)
        )
    }
}
